iceoryx_utils/include/iceoryx_utils/internal/lockfree_queue/cyclic_index.hpp

```diff
@@ -0,0 +1,117 @@
+// which might require minor adaptations/extensions of the CyclicIndex
+
+/// @brief index structure that can contain logical values 0, ..., CycleLength-1
+/// but also stores an internal cycle counter to be used in compare_exchange
+template <uint64_t CycleLength, typename ValueType = uint64_t>
+class CyclicIndex
+{
+  public:
+    using value_t = ValueType;
+
+    static constexpr ValueType MAX_INDEX = CycleLength - 1;
+    static constexpr ValueType MAX_VALUE = std::numeric_limits<ValueType>::max();
+
+    // assumes MAX_VALUE >= CycleLength, otherwise we could not fit in even one cycle
+    static constexpr ValueType MAX_CYCLE = MAX_VALUE / CycleLength;
+
+    static constexpr ValueType INDEX_AT_MAX_VALUE = MAX_VALUE % CycleLength;
+    static constexpr ValueType OVERFLOW_START_INDEX = (INDEX_AT_MAX_VALUE + 1) % CycleLength;
+
+    static_assert(CycleLength < MAX_VALUE / 2); // need at least one bit for the cycle
+    static_assert(CycleLength > 0);
+
+    explicit CyclicIndex(ValueType value = 0) noexcept
+        : m_value(value)
+    {
+    }
+
+    CyclicIndex(ValueType index, ValueType cycle) noexcept
+        : m_value(index + cycle * CycleLength)
+    {
+    }
+
+    CyclicIndex(const CyclicIndex&) = default;
+    CyclicIndex& operator=(const CyclicIndex&) = default;
+
+    ValueType getIndex() const noexcept
+    {
+        return m_value % CycleLength;
+    }
+
+    ValueType getCycle() const noexcept
+    {
+        return m_value / CycleLength;
+    }
+
+    CyclicIndex operator+(ValueType value) const
+    {
+        // if we were at this value, we would have no overflow, i.e. when m_value is larger there is an overflow
+        auto delta = MAX_VALUE - value;
+        if (delta < m_value)
+        {
+            // overflow, rare case (overflow by m_value - delta)
+            // we need to compute the correct index and cycle we are in after overflow
+            // note that we could also limit the max value to always start at OVERFLOW_START_INDEX = 0,
+            // but this has other drawbacks (and the overflow will not occur often if at all with 64 bit)
+            delta = m_value - delta - 1;
+            return CyclicIndex(OVERFLOW_START_INDEX + delta);
+        }
+
+        // no overflow, regular case
+        return CyclicIndex(m_value + value);
+    }
+
+    CyclicIndex next()
+    {
+        if (m_value == MAX_VALUE)
+        {
+            return CyclicIndex(OVERFLOW_START_INDEX);
+        }
+        return CyclicIndex(m_value + 1);
+    }
+
+    bool isOneCycleBehind(const CyclicIndex& other)
+    {
+        auto thisCycle = this->getCycle();
+        auto otherCycle = other.getCycle();
+
+        if (thisCycle == MAX_CYCLE)
+        {
+            return otherCycle == 0;
+        }
+        return isBehind(thisCycle, otherCycle, 1);
+    }
+
+  private:
+    ValueType m_value{0};
+
+    static bool isBehind(ValueType cycle1, ValueType cycle2, ValueType cyclesBehind = 1)

```

Comment:
Why are we having here the extra parameter cyclesBehind? This method is only used once in isOnCycleBehind with the default argument 1. Since it is a one liner and with the context of isOnCycleBehind I would suggest of getting rid of this static function and move the line directly into isOnCycleBehind like:
```
return (thisCycle + 1 == otherCycle);
```
With this you even reduce your code in isOneCycleBehind ;)

I think this is overabstraction

Reply:
This is artifact from a while ago when there were more "combinations" of cycle and CyclicIndex object argument. There it reduced some duplication, but is not needed any more.